pkg/apis/pipeline/v1alpha1/taskrun_validation_test.go

```diff
@@ -0,0 +1,326 @@
+						ResourceRef: PipelineResourceRef{
+							Name: "testresource",
+						},
+					}},
+					Params: []Param{{
+						Name:  "name",
+						Value: "value",
+					}, {
+						Name:  "name",
+						Value: "value",
+					}},
+				},
+			},
+			wantErr: apis.ErrMultipleOneOf("spec.inputs.params"),
+		},
+		{
+			name: "invalid task output type",
+			spec: &TaskRunSpec{
+				TaskRef: TaskRef{
+					Name: "taskrefname",
+				},
+				Outputs: Outputs{
+					Resources: []TaskResource{{
+						Name: "resourceName",
+						Type: "invalidtype",
+					}},
+				},
+			},
+			wantErr: apis.ErrInvalidValue("invalidtype", "spec.Outputs.Resources.resourceName.Type"),
+		},
+		{
+			name: "duplicate task output name",
+			spec: &TaskRunSpec{
+				TaskRef: TaskRef{
+					Name: "taskrefname",
+				},
+				Outputs: Outputs{
+					Resources: []TaskResource{{
+						Type: "git",
+						Name: "resource1",
+					}, {
+						Type: "git",
+						Name: "resource1",
+					}},
+				},
+			},
+			wantErr: apis.ErrMultipleOneOf("spec.Outputs.Resources.Name"),
+		},
+		{
+			name: "valid task output",
+			spec: &TaskRunSpec{
+				TaskRef: TaskRef{
+					Name: "taskrefname",
+				},
+				Outputs: Outputs{
+					Resources: []TaskResource{{
+						Type: "git",
+					}},
+				},
+			},
+		},
+		{
+			name: "invalid task type result",
+			spec: &TaskRunSpec{
+				TaskRef: TaskRef{
+					Name: "taskrefname",
+				},
+				Results: Results{
+					Logs: ResultTarget{
+						Name: "resultlogs",
+						Type: "wrongtype",
+					},
+				},
+			},
+			wantErr: apis.ErrInvalidValue("wrongtype", "spec.results.logs.Type"),
+		},
+		{
+			name: "invalid task type result",
+			spec: &TaskRunSpec{
+				TaskRef: TaskRef{
+					Name: "taskrefname",
+				},
+				Results: Results{
+					Runs: ResultTarget{
+						Name: "resultlogs",
+						Type: "wrongtype",
+					},
+				},
+			},
+			wantErr: apis.ErrInvalidValue("wrongtype", "spec.results.runs.Type"),
+		},
+		{
+			name: "invalid task type result",
+			spec: &TaskRunSpec{
+				TaskRef: TaskRef{
+					Name: "taskrefname",
+				},
+				Results: Results{
+					Tests: &ResultTarget{
+						Name: "resultlogs",
+						Type: "wrongtype",
+					},
+				},
+			},
+			wantErr: apis.ErrInvalidValue("wrongtype", "spec.results.tests.Type"),
+		},
+		{
+			name: "invalid task type result name",
+			spec: &TaskRunSpec{
+				TaskRef: TaskRef{
+					Name: "taskrefname",
+				},
+				Results: Results{
+					Runs: ResultTarget{
+						Name: "",
+						Type: ResultTargetTypeGCS,
+					},
+				},
+			},
+			wantErr: apis.ErrMissingField("spec.results.runs.name"),
+		},
+		{
+			name: "invalid task type result url",
+			spec: &TaskRunSpec{
+				TaskRef: TaskRef{
+					Name: "taskrefname",
+				},
+				Results: Results{
+					Runs: ResultTarget{
+						Name: "resultrunname",
+						Type: ResultTargetTypeGCS,
+						URL:  "",
+					},
+				},
+			},
+			wantErr: apis.ErrMissingField("spec.results.runs.URL"),
+		},
+		{
+			name: "valid task type result",
```

Comment:
one idea to reduced the boilerplate in the testcases would be to test validation of the individual types in separate tests, e.g. have a set of tests just for `Results`, one just for `Inputs`, one just for `Params`, etc., then instead of having to instantiate an entire `TaskRun` each time, you could just instantiate the section you need to validate

Reply:
I did think about that but I am generally not a fan of testing private functions. Probably I could make `Validate` functions on each of the types and separate the cases. That might be lot easier to read.